### PR TITLE
Nodejs kafka tutorial - Run through - support for oauthbearer token

### PR DESCRIPTION
**Pull request by pacostas**
Merge this PR after:
* https://github.com/nodeshift-starters/reactive-example/pull/92

This PR:
* Adds instructions on how to configure access for your client in order consume/produce messages from a kafka instance
* Adds instructions on how to enable SASL/OAUTHBEARER authentication mechanism.

### CODE REVIEW

File: docs/kafka/nodejs-kafka/README.adoc
Comment (AedinC) [inline docs/kafka/nodejs-kafka/README.adoc:186]: Could we get rid of the instruction to use PLAIN altogether here? So then we could add something like this: In the `rhoas.env` file, set the SASL connection mechanism and the Kafka instance client credentials as shown in the following configuration. Replace the values with your own credential information. {product-kafka} supports the SASL/OAUTHBEARER mechanism for authentication, which is the recommended authentication mechanism to use.

KAFKA_HOST=<bootstrap_server>
RHOAS_SERVICE_ACCOUNT_CLIENT_ID=<client_id>
RHOAS_SERVICE_ACCOUNT_CLIENT_SECRET=<client_secret>
KAFKA_SASL_MECHANISM=oauthbearer
RHOAS_TOKEN_ENDPOINT_URL=https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token

File: docs/kafka/nodejs-kafka/README.adoc
Comment (AedinC) [inline docs/kafka/nodejs-kafka/README.adoc:251]:
```suggestion
image::consumer-permissions-for-kafka-topic.png[Image of access configuration details for managing consumer permissions]
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (AedinC) [inline docs/kafka/nodejs-kafka/README.adoc:245]:
```suggestion
image::kafka-topic-manage-access.png[Image of access configuration details for managing access of kafka topic]
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (AedinC) [inline docs/kafka/nodejs-kafka/README.adoc:260]:
```suggestion
image::producer-permissions-for-kafka-topic.png[Image of access configuration details for managing producer permissions]
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (AedinC) [inline docs/kafka/nodejs-kafka/README.adoc:253]:
```suggestion
.. In the first list for *Topic* select `Starts with` and in the second list, select the `countries` topic.
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (AedinC) [inline docs/kafka/nodejs-kafka/README.adoc:254]:
```suggestion
.. In the first list of *Consumer group* select `Starts with` and for the second field, enter the value `consumer-test`.
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (AedinC) [inline docs/kafka/nodejs-kafka/README.adoc:262]:
```suggestion
..  In the first list for *Topic* select `Starts with` and in the second list, select the `countries` topic.
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (AedinC) [inline docs/kafka/nodejs-kafka/README.adoc:263]:
```suggestion
. Click *Save*.
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (AedinC) [inline docs/kafka/nodejs-kafka/README.adoc:264]:
```suggestion
.Verification
ifdef::qs[]
* Are the accounts listed on the *Access* page?
endif::[]
ifndef::qs[]
* Verify that the accounts are listed on the *Access* page.
endif::[]
```

**Review by AedinC — CHANGES_REQUESTED** [review]
Thanks for this Costas. I've made some suggestions. :)

File: docs/kafka/nodejs-kafka/README.adoc
Comment (mhdawson) [reply docs/kafka/nodejs-kafka/README.adoc:186]: For the tutorial I'm happy to only mention OAUTHBEARER, even though the code will support both.

Comment (pacostas) [reply docs/kafka/nodejs-kafka/README.adoc:186]: @AedinC Is it ok to push it in a commit ? and have a review on that too? Practically i'll copy paste the above text you wrote and replace the `plain` instructions.
```
Could we get rid of the instruction to use PLAIN altogether here? So then we could add something like this: In the rhoas.env file, set the SASL connection mechanism and the Kafka instance client credentials as shown in the following configuration. Replace the values with your own credential information. {product-kafka} supports the SASL/OAUTHBEARER mechanism for authentication, which is the recommended authentication mechanism to use.

KAFKA_HOST=<bootstrap_server>
RHOAS_SERVICE_ACCOUNT_CLIENT_ID=<client_id>
RHOAS_SERVICE_ACCOUNT_CLIENT_SECRET=<client_secret>
KAFKA_SASL_MECHANISM=oauthbearer
RHOAS_TOKEN_ENDPOINT_URL=https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token
```

Comment (jbyrne-redhat) [reply docs/kafka/nodejs-kafka/README.adoc:186]: @pacostas Aedin is on PTO this week, so I'll do another review of the updates for you :+1: . I actually worked with @mhdawson on the original version of this quick start.

Comment (pacostas) [reply docs/kafka/nodejs-kafka/README.adoc]: Its a very interesting tutorial and a hot (not kafka) topic these days :p  
@jbyrne-redhat Here is the commit I just pushed, with the text pasted from Aedin https://github.com/redhat-developer/app-services-guides/pull/600/commits/73c2264fc704206b5568ffcc016a4d014f77b9ba

File: docs/kafka/nodejs-kafka/README.adoc
Comment (jbyrne-redhat) [inline docs/kafka/nodejs-kafka/README.adoc:101]: @pacostas Just curious why we removed the noted limitation about node-rdkafka in addition to changing the minimum specified version of Node.js. Is that limitation no longer true?

File: docs/kafka/nodejs-kafka/README.adoc
Comment (jbyrne-redhat) [inline docs/kafka/nodejs-kafka/README.adoc:164]: Suggest a slight rewording here:

```
. In the `rhoas.env` file, set the SASL authentication mechanism and the Kafka instance client credentials as shown in the following configuration. Replace the values with your own credential information. The configuration uses the SASL/OAUTHBEARER authentication mechanism, which is the recommended authentication mechanism to use in {product-kafka}.
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (jbyrne-redhat) [inline docs/kafka/nodejs-kafka/README.adoc:173]: There is an available attribute for the token endpoint URL. Therefore:

```
.Setting environment variables in the rhoas.env file
[source,subs="+attributes,+quotes"]
----
KAFKA_HOST=__<bootstrap_server>__
RHOAS_SERVICE_ACCOUNT_CLIENT_ID=__<client_id>__
RHOAS_SERVICE_ACCOUNT_CLIENT_SECRET=__<client_secret>__
KAFKA_SASL_MECHANISM=plain
KAFKA_SASL_MECHANISM=oauthbearer
RHOAS_TOKEN_ENDPOINT_URL={sso-token-url}
```

Comment (jbyrne-redhat) [reply docs/kafka/nodejs-kafka/README.adoc:173]: As part of the change suggested above, observe that we updated the `subs` element at the head of the code block to also include attributes.

File: docs/kafka/nodejs-kafka/README.adoc
Comment (jbyrne-redhat) [inline docs/kafka/nodejs-kafka/README.adoc:228]: @pacostas Thanks so much for adding this detailed procedure on setting access permissions for the `countries` topic.

With that said, I'm not sure this is _quite_ the approach we want to follow here. In general, we continue to add more and more examples and guides for connecting different types of clients to OpenShift Streams. Those examples all use topics with various names and require different access permissions. It's not really scalable for us to write a tailored procedure each time for setting access permissions. It also just increases the number of procedures that we must separately maintain and verify in our documentation set.

We have a [detailed guide](https://access.redhat.com/documentation/en-us/red_hat_openshift_streams_for_apache_kafka/1/guide/2f4bf7cf-5de2-4254-8274-6bf71673f407#_6b9cf9a6-5bbf-43ab-8ee6-b9f33630bf64) on setting and managing access permissions, and that should be our single point of reference as much as possible in our documentation set. That guide includes example scenarios for setting produce and consume permissions for a specified topic.

Therefore, the changes I recommend here are the following:

- We omit this dedicated procedure on configuring access permissions for the `countries` topic
- In the final `Running the Node.js example application` section, we add one more prerequisite, stating that you must have configured access permissions for the `countries` topic. That should include a link our [access management guide](https://access.redhat.com/documentation/en-us/red_hat_openshift_streams_for_apache_kafka/1/guide/2f4bf7cf-5de2-4254-8274-6bf71673f407#_6b9cf9a6-5bbf-43ab-8ee6-b9f33630bf64). I will make that suggestion on the relevant line below.

Comment (jbyrne-redhat) [reply docs/kafka/nodejs-kafka/README.adoc:228]: Based on the new prerequisite I've suggested above, we should also remove the final prerequisite in the `Configuring the Node.js example application to connect to a Kafka instance` section (line 157). That prereq is not actually needed at that point in the example anyway.

File: docs/kafka/nodejs-kafka/README.adoc
Comment (pacostas) [reply docs/kafka/nodejs-kafka/README.adoc:101]: @jbyrne-redhat I changed it from 14 to 16 because on May will go EOL, although I agree keeping it to 14. I think it would be nice to add something like `Node.js 14 or later is installed.` if you think this would help.
Yes, this limitation no longer is true, node-rdkafka can run on node versions 14, 16 and 18.

File: docs/kafka/nodejs-kafka/README.adoc
Comment (pacostas) [reply docs/kafka/nodejs-kafka/README.adoc:164]: @jbyrne-redhat can you approve?

File: docs/kafka/nodejs-kafka/README.adoc
Comment (pacostas) [inline docs/kafka/nodejs-kafka/README.adoc:166]:
```suggestion
.Setting environment variables in the rhoas.env file [source,subs="+attributes,+quotes"] ---- KAFKA_HOST=__<bootstrap_server>__ RHOAS_SERVICE_ACCOUNT_CLIENT_ID=__<client_id>__ RHOAS_SERVICE_ACCOUNT_CLIENT_SECRET=__<client_secret>__ KAFKA_SASL_MECHANISM=plain KAFKA_SASL_MECHANISM=oauthbearer RHOAS_TOKEN_ENDPOINT_URL={sso-token-url}
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (pacostas) [inline docs/kafka/nodejs-kafka/README.adoc:167]:
```suggestion
[source,subs="+attributes,+quotes"]
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (jbyrne-redhat) [reply docs/kafka/nodejs-kafka/README.adoc:164]: :+1:

File: docs/kafka/nodejs-kafka/README.adoc
Comment (pacostas) [inline docs/kafka/nodejs-kafka/README.adoc:173]:
```suggestion
RHOAS_TOKEN_ENDPOINT_URL={sso-token-url}
```

**Review by pacostas — COMMENTED** [review]
@jbyrne-redhat Can you approve?

File: docs/kafka/nodejs-kafka/README.adoc
Comment (jbyrne-redhat) [reply docs/kafka/nodejs-kafka/README.adoc:101]: Yes, I agree:

```
* https://nodejs.org/en/download/[Node.js 14^] or later is installed
```

File: docs/kafka/nodejs-kafka/README.adoc
Comment (pacostas) [reply docs/kafka/nodejs-kafka/README.adoc]: I like when everything is included in a tutorial and in case I want to dig further I can follow the references. Mostly I saw it as a friction point, where despite following all the steps on the tutorial, my setup didnt work, and I had to spend time to find some extra pieces.

Although, I totally understand, duplication is not maintainable and should be avoided. I've added your suggestions under this commit https://github.com/redhat-developer/app-services-guides/pull/600/commits/5788cfae2eadac69d06b9adbe7c7d8be9d5af070

@jbyrne-redhat can you approve?

File: docs/kafka/nodejs-kafka/README.adoc
Comment (pacostas) [inline docs/kafka/nodejs-kafka/README.adoc:101]:
```suggestion
* https://nodejs.org/en/download/[Node.js 14^] or later is installed.
```

**Review by jbyrne-redhat — APPROVED** [review]
I previewed the HTML and these updates all look good to me, @pacostas. Thank you again for making these :)

**Review by AedinC — APPROVED** [review]

**Comment by mhdawson** [comment]
**NOTE:** This should wait to land until we get the sample updated - https://github.com/nodeshift-starters/reactive-example/pull/92

**Comment by pacostas** [comment]
@AedinC The PR https://github.com/nodeshift-starters/reactive-example/pull/92 has been merged :)

**Comment by mhdawson** [comment]
@pacostas many thanks for helping to get the started updated and this into a good shape.

**Comment by AedinC** [comment]
> @AedinC The PR [nodeshift-starters/reactive-example#92](https://github.com/nodeshift-starters/reactive-example/pull/92) has been merged :)

Thanks so much for all your work on this @pacostas.

**Comment by AedinC** [comment]
> @AedinC The PR [nodeshift-starters/reactive-example#92](https://github.com/nodeshift-starters/reactive-example/pull/92) has been merged :)

Hi @pacostas. Do you want me to merge this PR? (Not used to merging PRs that aren't mine. :) )

**Comment by pacostas** [comment]
>

**Comment by pacostas** [comment]
> > @AedinC The PR [nodeshift-starters/reactive-example#92](https://github.com/nodeshift-starters/reactive-example/pull/92) has been merged :)
> 
> Hi @pacostas. Do you want me to merge this PR? (Not used to merging PRs that aren't mine. :) )

I cant merge this PR on this Repo, I dont have access write access. I can only open PR :)

**Comment by pacostas** [comment]
BTW accidentaly closed and re-opened this PR, can you approve the running workflows once more? @AedinC

**Comment by AedinC** [comment]
> BTW accidentaly closed and re-opened this PR, can you approve the running workflows once more? @AedinC

Done. Will merge now. Thanks again @pacostas!

**Pull request merged**
